tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs

```diff
@@ -0,0 +1,234 @@
+
+        private readonly List<KeyValuePair<string, string>> _tags;
+        private string? _cachedPropagationHeader;
+
+        public TraceTagCollection()
+        {
+            _tags = new List<KeyValuePair<string, string>>(DefaultCapacity);
+        }
+
+        private TraceTagCollection(List<KeyValuePair<string, string>> tags)
+        {
+            _tags = tags;
+        }
+
+        public static TraceTagCollection ParseFromPropagationHeader(string? propagationHeader)
+        {
+            if (string.IsNullOrEmpty(propagationHeader))
+            {
+                return new TraceTagCollection();
+            }
+
+            var tags = propagationHeader!.Split(TagPairSeparators, StringSplitOptions.RemoveEmptyEntries);
+            var tagList = new List<KeyValuePair<string, string>>(tags.Length);
+            var cacheOriginalHeader = true;
+
+            foreach (var tag in tags)
+            {
+                // the shortest tag has the "_dd.p." prefix, a 1-character key, and 1-character value (e.g. "_dd.p.a=b")
+                if (tag.Length >= MinimumPropagationHeaderLength)
+                {
+                    // NOTE: the first equals sign is the separator between key/value, but the tag value can contain
+                    // additional equals signs, so make sure we only split on the _first_ one. For example,
+                    // the "_dd.p.upstream_services" tag will have base64-encoded strings which use '=' for padding.
+                    var separatorIndex = tag.IndexOf(KeyValueSeparator);
+
+                    // "_dd.p.a=b"
+                    //         ⬆   separator must be at index 7 or higher and before the end of string
+                    //  012345678
+                    if (separatorIndex > PropagatedTagPrefixLength &&
+                        separatorIndex < tag.Length - 1 &&
+                        tag.StartsWith(PropagatedTagPrefix, StringComparison.Ordinal))
+                    {
+                        var key = tag.Substring(0, separatorIndex);
+                        var value = tag.Substring(separatorIndex + 1);
+                        tagList.Add(new KeyValuePair<string, string>(key, value));
+                    }
+                    else
+                    {
+                        // skip invalid tag
+                        cacheOriginalHeader = false;
+                    }
+                }
+                else
+                {
+                    // skip invalid tag
+                    cacheOriginalHeader = false;
+                }
+            }
+
+            var traceTags = new TraceTagCollection(tagList);
+
+            if (cacheOriginalHeader)
+            {
+                // we didn't skip any invalid tag, we can cache the original header string
+                traceTags._cachedPropagationHeader = propagationHeader;
+            }
+
+            return traceTags;
+        }
+
+        public List<KeyValuePair<string, string>> AsList() => _tags;
```

Comment:
The reason I exposed the internal `List<T>` instead of implementing `IEnumerable<KVP<string,string>>` was so we can iterate over it without the extra heap allocation. It's not great because it let's callers modify the list and skip all the validation, etc.

That's probably premature optimization on my part. I'll change `TraceTagCollection` to implement `IEnumerable<KVP<string,string>>` for now.

Reply:
I think the performant accepted way is having a custom struct base enumerator to avoid extra allocations if this is being use in a hot path

Reply:
Actually, I was able to return the list's own struct enumerator, so best of both worlds?
- 9f602c4a8bcd7197fc9d26b7016eca8dbe113186 implement `IEnumerable<KVP<string,string>>`
- 41004cfd063dd187e527721f5bc3079703496d00 enumerate the collection directly
- 59ff634c59bf79d471ede2d42d8ff6dddb5a58f1 remove unused `AsList()`

Reply:
(didn't see Tony's comment until I refreshed)
That's exactly what I did, @tonyredondo. Thanks!